core/src/banking_stage/transaction_scheduler/multi_iterator_consume_scheduler.rs

```diff
@@ -0,0 +1,617 @@
+                let ids = work.ids.clone();
+                (work, ids)
+            })
+            .unzip()
+    }
+
+    #[test]
+    fn test_schedule_disconnected_channel() {
+        let (mut scheduler, work_receivers) = create_test_frame(1);
+        let mut container = create_container([(&Keypair::new(), &[Pubkey::new_unique()], 1, 1)]);
+
+        drop(work_receivers); // explicitly drop receivers
+        assert_matches!(
+            scheduler.schedule(&mut container),
+            Err(SchedulerError::DisconnectedSendChannel(_))
+        );
+    }
+
+    #[test]
+    fn test_schedule_single_threaded_no_conflicts() {
+        let (mut scheduler, work_receivers) = create_test_frame(1);
+        let mut container = create_container([
+            (&Keypair::new(), &[Pubkey::new_unique()], 1, 1),
+            (&Keypair::new(), &[Pubkey::new_unique()], 2, 2),
+        ]);
+
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+        assert_eq!(collect_work(&work_receivers[0]).1, vec![txids!([1, 0])]);
+    }
+
+    #[test]
+    fn test_schedule_single_threaded_conflict() {
+        let (mut scheduler, work_receivers) = create_test_frame(1);
+        let pubkey = Pubkey::new_unique();
+        let mut container = create_container([
+            (&Keypair::new(), &[pubkey], 1, 1),
+            (&Keypair::new(), &[pubkey], 1, 2),
+        ]);
+
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 2);
+        assert_eq!(
+            collect_work(&work_receivers[0]).1,
+            vec![txids!([1]), txids!([0])]
+        );
+    }
+
+    #[test]
+    fn test_schedule_consume_single_threaded_multi_batch() {
+        let (mut scheduler, work_receivers) = create_test_frame(1);
+        let mut container = create_container(
+            (0..4 * TARGET_NUM_TRANSACTIONS_PER_BATCH)
+                .map(|i| (Keypair::new(), [Pubkey::new_unique()], i as u64, 1)),
+        );
+
+        // expect 4 full batches to be scheduled
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 4 * TARGET_NUM_TRANSACTIONS_PER_BATCH);
+
+        let thread0_work_counts: Vec<_> = work_receivers[0]
+            .try_iter()
+            .map(|work| work.ids.len())
+            .collect();
+        assert_eq!(thread0_work_counts, [TARGET_NUM_TRANSACTIONS_PER_BATCH; 4]);
+    }
+
+    #[test]
+    fn test_schedule_simple_thread_selection() {
+        let (mut scheduler, work_receivers) = create_test_frame(2);
+        let mut container =
+            create_container((0..4).map(|i| (Keypair::new(), [Pubkey::new_unique()], 1, i)));
+
+        let num_scheduled = scheduler.schedule(&mut container).unwrap();
+        assert_eq!(num_scheduled, 4);
+        assert_eq!(collect_work(&work_receivers[0]).1, [txids!([3, 1])]);
+        assert_eq!(collect_work(&work_receivers[1]).1, [txids!([2, 0])]);
```

Comment:
Let's say we have `N` *distinct* markets near the top of our queue.


Conflict graph might look something like this:
```mermaid
graph LR;

A1((A1)) --> A2((A2)) --> A3((A3)) --> A4((..))
B1((B1)) --> B2((B2)) --> B3((B3)) --> B4((..))
C1((C1)) --> C2((C2)) --> C3((C3)) --> C4((..))
D1((D1)) --> D2((D2)) --> D3((D3)) --> D4((..))
```

This view makes it somewhat clear a reasonable choice is to have A txs on one thread, B txs on another, and so on.

But the priority queue might look like this:
```mermaid
graph LR;
A1((A1));
A2((A2));
A3((A3));
A4((A4));

B1((B1));
B2((B2));
B3((B3));
B4((B4));

C1((C1));
C2((C2));
C3((C3));
C4((C4));

D1((D1));
D2((D2));
D3((D3));
D4((D4));

A1 --> A2 --> B1 --> C1 --> B2 --> D1 --> B3 --> A3 --> D2 --> C2 --> C3 --> A4 --> B4 --> C4 --> D3 --> D4
```

I'll just walk through the first few steps here, and maybe that will clear it up. Assume we start with a completely empty schedule:
1. See `A1` - can be scheduled on any thread. Choose thread with fewest scheduled. Send to thread 0.
2. See `A2` - skip for this inner-iteration since it conflicts with `A1`
3. See `B1` - can be scheduled on any thread. Choose thread with fewest scheduled. Send to thread 1.
4. See `C1` - can be scheduled on any thread. Choose thread with fewest scheduled. Send to thread 2.
5. See `B2` - skip for this inner-iteration since it conflicts with `B1`
6. See `D1` - can be scheduled on any thread. Choose thread with fewest scheduled. Send to thread 3.
7. Remaining iterators reach end of queue since there's no more txs that don't conflict. 4 iterators are at A1, B1, C1, D1.
8. See `A2` - can only be scheduled on thread 0. Send to thread 0.
9. See `B2` - can only be scheduled on thread 1. Send to thread 1.

And so on. By round-robining non-conflicting txs at the beginning of our MI, we are effectively distributing the distinct top-of-queue conflict chains into our threads.

Obviously this is not perfect, since if we have non-conflicting txs at top-of-queue, but the txs have shared lower-priority conflicts (i.e. lower priority that conflict with both `A1` and `B1` for example) will cause later blocks where we have to wait for txs to finish.